### ISSUE TITLE
stable API to deal with "fake" SCI types like records

### ISSUE THREAD

**Issue by borkdude** [babashka/sci#753]
Conversation with @frenchy64:

@frenchy64 wrote:

It feels like this abstraction could be a bit cleaner. it seems too tied to records, I think I want an abstraction over

- "is this a fake class/prototype created due to limitations of sci" and

- "what is the name of this class/prototype, if any?".

Could we reify these metadata symbols as records themselves? like provide a sci.runtime.Type record that responds to a type-name fn that returns a nilable string. 

That way if you decide to fake classes for other things (definterface, defprotocol, deftype etc), you can build on top of Type without requiring changes in schema. and anywhere you normally do (extend-protocol ... Class) you can do (extend-protocol ... sci.runtime.Type) also, without clashing with existing (extend-protocol ... Symbol). 

My response:

We already have a `sci.lang` namespace like this:

```
(ns sci.lang
  {:no-doc true})

;; marker interface for vars, clj only for now
#?(:clj (definterface IVar))
```

This "class" is also exposed within SCI, so you can extend protocols to that interface.

We could add more marker interfaces (or in CLJS protocols) there and then have a `deftype` which implements these.

E.g:

```
(defprotocol IType (type-name [_]))
```

and a fake record type could then be:

```
(deftype SciRecordType [..]
  sci.lang.IType
  (type-name [..] ..)
```

and `sci.lang.IType` + `sci.lang/type-name` would then be exposed in the SCI runtime.

**Comment by borkdude** [comment]
I guess we could :)

**Comment by frenchy64** [comment]
Thinking of corner cases: is there a canonical version of each sci class? ie., is `(identical? Foo Foo)` or `(= Foo Foo)`? I don't think I saw a unit test for this anyhow.

**Comment by borkdude** [comment]
Yes, there should be only one instance, but will add tests for that.

**Comment by borkdude** [comment]
Fixed with 05727e9ddccc9d641d1d4ba05cd444904ec66514
